redis/src/main/scala/akka/streams/alpakka/redis/impl/RedisStreamingSourceStage.scala

```diff
@@ -0,0 +1,174 @@
+          } else if (shouldTerminateStage) {
+            completeStage()
+          }
+      }
+    )
+
+    override def postStop(): Unit =
+      super.postStop()
+
+    def handleSuccess(result: RedisKeyValue[K, V]): Unit = {
+
+      currentRequests = currentRequests - 1
+
+      (buffer.isEmpty, isAvailable(out)) match {
+        case (false, true) =>
+          push(out, buffer.poll())
+          buffer.offer(result)
+        case (true, true) =>
+          push(out, result)
+        case (_, false) =>
+          buffer.offer(result)
+      }
+      if (shouldTerminateStage) {
+        completeStage()
+      }
+
+    }
+
+    private def shouldTerminateStage: Boolean =
+      redisFutureCompleted &&
+      currentRequests == 0 &&
+      buffer.isEmpty
+
+    def checkConcurrency(): Boolean = {
+      currentRequests = currentRequests + 1
+      maxBufferSize > buffer.size &&
+      maxConcurrency > currentRequests
+    }
+
+    def handleFailure(ex: Exception): Unit =
+      failStage(ex)
+
+  }
+
+  override def shape: SourceShape[RedisKeyValue[K, V]] = SourceShape.of(out)
+}
+
+/**
+ * Internal API
+ */
+@InternalApi
+private[redis] class RedishmgetSourceStage[K, V](key: K,
+                                                 fields: Seq[K],
+                                                 val connection: StatefulRedisConnection[K, V],
+                                                 val settings: RedisSubscriberSettings =
+                                                   RedisSubscriberSettings.Defaults)
+    extends RedisStreamingSourceStage[K, V] {
+
+  def resultHandler(checkConcurrency: () => Boolean,
+                    successCallback: AsyncCallback[RedisKeyValue[K, V]],
+                    failureCallback: AsyncCallback[Exception]): RedisFuture[lang.Long] =
+    connection
+      .async()
+      .hmget(
+        new KeyValueStreamingChannel[K, V]() {
+          override def onKeyValue(key: K, value: V): Unit =
+            if (checkConcurrency()) {
+              successCallback.invoke(RedisKeyValue(key, value))
+            } else {
+              failureCallback.invoke(BufferOverflowException("Redis Source buffer overflown"))
+            }
+        },
+        key,
+        fields: _*
+      )
+
+}
+
+/**
+ * Internal API
+ */
+@InternalApi
+private[redis] class RedishgetallSourceStage[K, V](val key: K,
+                                                   override val connection: StatefulRedisConnection[K, V],
+                                                   override val settings: RedisSubscriberSettings =
+                                                     RedisSubscriberSettings.Defaults)
+    extends RedisStreamingSourceStage[K, V] {
+
+  def resultHandler(checkConcurrency: () => Boolean,
+                    successCallback: AsyncCallback[RedisKeyValue[K, V]],
+                    failureCallback: AsyncCallback[Exception]): RedisFuture[lang.Long] =
+    connection
+      .async()
+      .hgetall(
```

Comment:
`hgetall` can be adapted to akka-stream `Source` directly using lettuce reactive client:

```scala
import scala.collection.JavaConverters._
val source: Source[KeyValue[K, V], NotUsed] =
  Source
    .fromPublisher(connection.reactive().hgetall(key))
    .mapConcat(_.entrySet().asScala.toIndexedSeq)
    .map(entry => KeyValue.just(entry.getKey, entry.getValue))
```